### DISCUSSION TITLE
Problematic rewrite rules in the optimizer.

### DISCUSSION THREAD

**Discussion by yaotech2023** [apache/doris#17449] [RFC]
Hi, I'm browsing the optimizer code of doris, and I found that there may be some bugs in rules like InApplyToJoin and ScalarApplyToJoin because the null propagation in the expressions is not considered when trying to transform the **Apply** to **Join**.

**Comment by yaotech2023** [comment]
Think about the queries below:

SELECT * FROM t1 WHERE c2 IN (SELECT COALESCE(c2,0) FROM t2)

SELECT * FROM t1 LEFT SEMI JOIN t2 ON t1.c2 = COALESCE(c2,0)

They are not equal.

**Comment by englefly** [comment]
COALESCE  is special. it should be handled like an aggregate function.
another case is bitmap

**Reply by yaotech2023** [reply]
Well, I just tried to give an example about this situation. In fact you can change it to a lot of similar expressions like below:
select 1 from t2
select case c2 when not null then c2 else 0 from t2
....

**Comment by morrySnow** [comment]
it couldn't be better that describe more details: the data in t1, data in t2, the in predicate result and semi join result

**Comment by yaotech2023** [comment]
Well, there is an solution for this problem, but I'm not sure if it is a good way for doris.
